[dem] local coordination number implementation

maceligueta opened the pull request:
The coordination number can be achieved in two ways:
- local growth
- global growth
However, a single particle can't get 5.2 neighbours, only 4,5 or 6 (ints). So, if local growth is used, it is mandatory to have global growth activated as well, for a final refined growth of search distance.
Both local and global ways of approaching the coordination number can be activated or deactivated from the json file.

Commit:
local coordination number implementation

philbucher commented on applications/DEMApplication/custom_strategies/strategies/explicit_solver_continuum.cpp:
```diff
@@ -399,65 +399,125 @@ namespace Kratos {
+
+        if (desired_coordination_number <= 0.0) {
+            KRATOS_THROW_ERROR(std::runtime_error, "The specified Coordination Number is less or equal to zero, N.C. = ", desired_coordination_number)
+        }
```
```suggestion
        KRATOS_ERROR_IF(desired_coordination_number <= 0.0) << "The specified Coordination Number is less or equal to zero, N.C. = " << desired_coordination_number << std::endl;
```

philbucher commented on applications/DEMApplication/custom_strategies/strategies/explicit_solver_continuum.cpp:
```diff
@@ -399,65 +399,125 @@ namespace Kratos {
-            KRATOS_WARNING("DEM") << "Setting up Coordination Number by increasing or decreasing the search radius... " << std::endl;
-        }
+            KRATOS_INFO("DEM") << "Setting up Coordination Number (input = "<<desired_coordination_number<<") by increasing or decreasing the search radius. ";
+            if(local_coordination_option) KRATOS_INFO("") << "Local extension activated. ";
```
```suggestion
           KRATOS_INFO_IF("", local_coordination_option) KRATOS_INFO("") << "Local extension activated." << std::endl;
```

philbucher commented on applications/DEMApplication/custom_strategies/strategies/explicit_solver_continuum.cpp:
```diff
@@ -399,65 +399,125 @@ namespace Kratos {
-        }
+            KRATOS_INFO("DEM") << "Setting up Coordination Number (input = "<<desired_coordination_number<<") by increasing or decreasing the search radius. ";
+            if(local_coordination_option) KRATOS_INFO("") << "Local extension activated. ";
+            if(global_coordination_option) KRATOS_INFO("") << "Global extension activated. ";
```
same

maceligueta replied:
done

philbucher commented on applications/DEMApplication/custom_strategies/strategies/explicit_solver_continuum.cpp:
```diff
@@ -399,65 +399,125 @@ namespace Kratos {
+                iteration++;
+                if (current_coordination_number == 0.0) {
+                    KRATOS_WARNING("DEM") << "Coordination Number method not supported in this case" << "\n" << std::endl;
+                    KRATOS_THROW_ERROR(std::runtime_error, "The specified tangency method is not supported for this problem, please use absolute value instead", " ")
```
better use `KRATOS_ERROR`

maceligueta replied:
done

philbucher commented on applications/DEMApplication/custom_strategies/strategies/explicit_solver_continuum.cpp:
```diff
@@ -399,65 +399,125 @@ namespace Kratos {
+        //STAGE 1, Local Coordination Number
+        double tolerance = 0.5;
+        double max_factor_between_iterations = 1.02;
+        double relative_error = 1000;
```
I guess those can be `const`

maceligueta replied:
They can't, they are overwritten some lines below.

philbucher commented on applications/DEMApplication/custom_strategies/strategies/explicit_solver_continuum.cpp:
```diff
@@ -399,65 +399,125 @@ namespace Kratos {
+                total_error.resize(mNumberOfThreads);
+
+                //#pragma omp parallel for
+                for (size_t i = 0; i < mListOfSphericContinuumParticles.size(); i++) {
```
```suggestion
                for (int i = 0; i < static_cast<int>(mListOfSphericContinuumParticles.size()); ++i) {
```

otherwise will not work in Win (if you enable omp)

philbucher commented on applications/DEMApplication/custom_strategies/strategies/explicit_solver_continuum.cpp:
```diff
@@ -399,65 +399,125 @@ namespace Kratos {
+
+                //#pragma omp parallel for
+                for (size_t i = 0; i < mListOfSphericContinuumParticles.size(); i++) {
+                    const size_t neighbour_elements_size = mListOfSphericContinuumParticles[i]->mNeighbourElements.size();
```
```suggestion
                    const std::size_t neighbour_elements_size = mListOfSphericContinuumParticles[i]->mNeighbourElements.size();
```

philbucher commented on applications/DEMApplication/custom_strategies/strategies/explicit_solver_continuum.cpp:
```diff
@@ -399,65 +399,125 @@ namespace Kratos {
-                KRATOS_THROW_ERROR(std::runtime_error, "The specified tangency method is not supported for this problem, please use absolute value instead", " ")
-                break;
-            }
+                std::vector<double> total_error;
```
I think this would be much easier if you use a reduction

e.g. like this:
https://github.com/KratosMultiphysics/Kratos/blob/6c6c37a10415592ca0cf2d47c315c64b25b9a9f6/applications/RANSApplication/custom_processes/auxiliary_processes/rans_nut_y_plus_wall_function_sensitivities_process.cpp#L71-L81

maceligueta replied:
Lol! Actually, I have had many bad experiences with openmp reductions. Usually with signed ints, getting totally unexpected results. @ipouplana  can confirm. I am so afraid of using them that I'd rather not....

philbucher replied:
oh really? Good to know
FYI @sunethwarna

philbucher commented on applications/DEMApplication/custom_strategies/strategies/explicit_solver_continuum.cpp:
```diff
@@ -506,6 +566,16 @@ namespace Kratos {
 
+    void ContinuumExplicitSolverStrategy::SetSearchRadiiOnAllParticles(ModelPart& r_model_part, const double added_search_distance, const double amplification) {
+        KRATOS_TRY
+        int number_of_elements = r_model_part.GetCommunicator().LocalMesh().ElementsArray().end() - r_model_part.GetCommunicator().LocalMesh().ElementsArray().begin();
```
```suggestion
        const int number_of_elements = r_model_part.GetCommunicator().LocalMesh().NumberOfElements();
```

philbucher commented on applications/DEMApplication/custom_strategies/strategies/explicit_solver_continuum.cpp:
```diff
@@ -399,65 +399,125 @@ namespace Kratos {
+                iteration++;
+                if (current_coordination_number == 0.0) {
+                    KRATOS_WARNING("DEM") << "Coordination Number method not supported in this case" << "\n" << std::endl;
+                    KRATOS_THROW_ERROR(std::runtime_error, "The specified tangency method is not supported for this problem, please use absolute value instead", " ")
```
better use `KRATOS_ERROR`

maceligueta replied:
done

philbucher commented on applications/DEMApplication/custom_strategies/strategies/explicit_solver_continuum.cpp:
```diff
@@ -399,65 +399,125 @@ namespace Kratos {
         added_search_distance = 0.0;
         double amplification = 1.0;
 
         if(r_model_part.GetCommunicator().MyPID() == 0) {
```
this check is no longer necessary, the logger only prints only on one rank, unless you use the `...ON_ALL_RANKS` logger macro

maceligueta replied:
ok, thanks

philbucher reviewed (commented):
some suggestions

Commit:
Update applications/DEMApplication/custom_strategies/strategies/explicit_solver_continuum.cpp

Co-Authored-By: Philipp Bucher <philipp.bucher@tum.de>

Commit:
Update applications/DEMApplication/custom_strategies/strategies/explicit_solver_continuum.cpp

Co-Authored-By: Philipp Bucher <philipp.bucher@tum.de>

Commit:
Update applications/DEMApplication/custom_strategies/strategies/explicit_solver_continuum.cpp

Co-Authored-By: Philipp Bucher <philipp.bucher@tum.de>

Commit:
Update applications/DEMApplication/custom_strategies/strategies/explicit_solver_continuum.cpp

Co-Authored-By: Philipp Bucher <philipp.bucher@tum.de>

Commit:
Update applications/DEMApplication/custom_strategies/strategies/explicit_solver_continuum.cpp

Co-Authored-By: Philipp Bucher <philipp.bucher@tum.de>

Commit:
KRATOS_INFO and KRATOS_ERROR

Commit:
Rank 1 is already detected by Kratos

maceligueta commented:
I just addressed all your comments @philbucher . Many thanks for taking your time in reading this. It is always good to dedicate a bit of time to update the code. Too much copy and paste this afternoon.
I will open an issue to replace all KRATOS_THROW_ERROR from the DEM.

salvalatorre reviewed (approved):
Let us try this!
